specs/networking/p2p-interface.md

```diff
@@ -0,0 +1,743 @@
+2. A specification of the three network interaction _domains_ of ETH2.0: (a) the gossip domain, (b) the discovery domain, \(c\) the Req/Resp domain.
+3. The rationale and further explanation for the design choices made in the previous two sections.
+4. An analysis of the maturity/state of the libp2p features required by this spec across the languages in which ETH 2.0 clients are being developed.
+
+## Table of Contents
+
+<!-- cmd: doctoc --maxlevel=2 p2p-interface.md -->
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+
+
+- [Network Fundamentals](#network-fundamentals)
+  - [Transport](#transport)
+  - [Encryption and identification](#encryption-and-identification)
+  - [Protocol Negotiation](#protocol-negotiation)
+  - [Multiplexing](#multiplexing)
+- [ETH2 network interaction domains](#eth2-network-interaction-domains)
+  - [Constants](#constants)
+  - [The gossip domain: gossipsub](#the-gossip-domain-gossipsub)
+  - [The discovery domain: discv5](#the-discovery-domain-discv5)
+  - [The Req/Resp domain](#the-reqresp-domain)
+- [Design Decision Rationale](#design-decision-rationale)
+  - [Transport](#transport-1)
+  - [Multiplexing](#multiplexing-1)
+  - [Protocol Negotiation](#protocol-negotiation-1)
+  - [Encryption](#encryption)
+  - [Gossipsub](#gossipsub)
+  - [Req/Resp](#reqresp)
+  - [Discovery](#discovery)
+  - [Compression/Encoding](#compressionencoding)
+- [libp2p Implementations Matrix](#libp2p-implementations-matrix)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+# Network Fundamentals
+
+This section outlines the specification for the networking stack in Ethereum 2.0 clients.
+
+Sections that have differing parameters for mainnet launch and interoperability testing are split into subsections. Sections that are not split have the same parameters for interoperability testing as mainnet launch.
+
+## Transport
+
+Even though libp2p is a multi-transport stack (designed to listen on multiple simultaneous transports and endpoints transparently), we hereby define a profile for basic interoperability.
+
+#### Interop
+
+All implementations MUST support the TCP libp2p transport, and it MUST be enabled for both dialing and listening (i.e. outbound and inbound connections).
+
+The libp2p TCP transport supports listening on IPv4 and IPv6 addresses (and on multiple simultaneously). Clients SHOULD allow the operator to configure the listen IP addresses and ports, including the addressing schemes (IPv4, IPv6).
+
+To facilitate connectivity, and avert possible IPv6 routability/support issues, clients participating in the interoperability testnet MUST expose at least ONE IPv4 endpoint.
+
+All listening endpoints must be publicly dialable, and thus not rely on libp2p circuit relay, AutoNAT or AutoRelay facilities.
```

Comment:
Why not? Is there some major blocker to dialing a node via circuit relay? 

We utilize circuit relay for our testnet orchestration, but we can hardcode inbound routes to meet this requirement if needed.

Reply:
IMO, relays add complexity at this stage, but should be eventually incorporated as they play an important role in standing up well interconnected network. 

Here are some questions I'd ask:

- how are relays discovered?
- what are the security implications?
- are regular peers supposed to provide this functionality or is it backed by standalone infra, or both?

Reply:
for testing purposes, and for those that don't have `auto*` features implemented, it's easier to require a straightforward networking setup to start with - these exotic extras are essentially workarounds for when networking is difficult - not a complication we necessarily want to have from the start, when testing things.

Reply:
Agree with the above. Let's add network topology/connectivity improvements/complexity incrementally.

Reply:
By phasing these aspects in, the barrier for WIP libp2p implementations to become _feature complete_ also becomes more lower and more attainable.

Reply:
Sounds good to me. We'll want relay support soon, but can work without it for interop testing.